File: dnn-providers/hip-kernel-provider/rocke/platform/cpp/core/dtypes.cpp
// Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
// SPDX-License-Identifier: MIT
/* Target-independent dtype names. Mirrors rocke.core.dtypes:
 * normalize_dtype -> rocke_normalize_dtype.
 * No architecture, IR builder, or lowering dependencies.
 */
#include "rocke/dtypes.h"

#include <string.h>

/* =========================================================================
 * target-independent dtype normalization
 * =========================================================================
 *
 * Mirrors core/dtypes.py::normalize_dtype: strip + lower(name), then
 * _DTYPE_ALIASES.get(key, key). The canonical RHS values are interned static
 * strings; unknown spellings pass through as the lowercased text in `lowered`.
 */

typedef struct rocke_dtype_alias
{
    const char* alias; /* lowercased key */
    const char* canonical; /* interned canonical value */
} rocke_dtype_alias_t;

/* Byte-for-byte the _DTYPE_ALIASES map (insertion order is irrelevant: lookup is
 * by exact lowercased key). */
static const rocke_dtype_alias_t k_dtype_aliases[] = {
    {"f16", "fp16"},
    {"half", "fp16"},
    {"fp16", "fp16"},
    {"bf16", "bf16"},
    {"bfloat16", "bf16"},
    {"f32", "fp32"},
    {"float", "fp32"},
    {"fp32", "fp32"},
    {"iu8", "iu8"},
    {"iu4", "iu4"},
    {"i8", "i8"},
    {"int8", "i8"},
    {"i4", "i4"},
    {"int4", "i4"},
    {"i32", ROCKE_DTYPE_I32},
    {"int32", ROCKE_DTYPE_I32},
    {"fp8", "fp8e4m3"},
    {"fp8e4m3", "fp8e4m3"},
    {"bf8", "bf8e5m2"},
    {"bf8e5m2", "bf8e5m2"},
    {"fp6", "fp6e2m3"},
    {"fp6e2m3", "fp6e2m3"},
    {"bf6", "fp6e3m2"},
    {"fp6e3m2", "fp6e3m2"},
    {"fp4", "fp4e2m1"},
    {"fp4e2m1", "fp4e2m1"},
};
#define K_NUM_DTYPE_ALIASES ((int)(sizeof(k_dtype_aliases) / sizeof(k_dtype_aliases[0])))

/* str.strip(): Python strips ASCII whitespace from both ends. */
static int rocke_dtype_is_ws(char c)
{
    return c == ' ' || c == '\t' || c == '\n' || c == '\r' || c == '\f' || c == '\v';
}

const char* rocke_normalize_dtype(const char* name, char* lowered, size_t cap)
{
    int i;
    size_t n;
    const char* start;
    const char* end;

    if(name == NULL)
    {
        /* Python would raise on None.strip(); the C contract requires a result.
         * Treat as empty string (no alias match, passes through as ""). */
        if(lowered != NULL && cap > 0)
        {
            lowered[0] = '\0';
        }
        return (lowered != NULL && cap > 0) ? lowered : "";
    }

    /* strip(): advance over leading/trailing whitespace */
    start = name;
    while(*start != '\0' && rocke_dtype_is_ws(*start))
    {
        start++;
    }
    end = start + strlen(start);
    while(end > start && rocke_dtype_is_ws(end[-1]))
    {
        end--;
    }

    /* lower() into the caller buffer (the pass-through result) */
    n = (size_t)(end - start);
    if(lowered == NULL || cap == 0)
    {
        /* No scratch: caller guarantees a known spelling. Build a small inline
         * copy on a fixed buffer to look up; if unknown we have nowhere to
         * return it, so fall back to "". */
        static char tmp[64];
        size_t m = n < sizeof(tmp) - 1 ? n : sizeof(tmp) - 1;
        for(i = 0; (size_t)i < m; i++)
        {
            char c = start[i];
            tmp[i] = (c >= 'A' && c <= 'Z') ? (char)(c - 'A' + 'a') : c;
        }
        tmp[m] = '\0';
        for(i = 0; i < K_NUM_DTYPE_ALIASES; i++)
        {
            if(strcmp(tmp, k_dtype_aliases[i].alias) == 0)
            {
                return k_dtype_aliases[i].canonical;
            }
        }
        return "";
    }

    if(n > cap - 1)
    {
        n = cap - 1;
    }
    for(i = 0; (size_t)i < n; i++)
    {
        char c = start[i];
        lowered[i] = (c >= 'A' && c <= 'Z') ? (char)(c - 'A' + 'a') : c;
    }
    lowered[n] = '\0';

    /* _DTYPE_ALIASES.get(key, key) */
    for(i = 0; i < K_NUM_DTYPE_ALIASES; i++)
    {
        if(strcmp(lowered, k_dtype_aliases[i].alias) == 0)
        {
            return k_dtype_aliases[i].canonical;
        }
    }
    return lowered;
}

File: dnn-providers/hip-kernel-provider/rocke/platform/python/rocke/core/dtypes.py
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
"""Target-independent dtype names shared by specs and architecture catalogs.

These names describe logical formats. Recognition does not imply an IR scalar
representation, conversion support, or a matrix instruction on a given target.
Target-specific support belongs to the architecture catalog.
"""

from __future__ import annotations

# Canonical dtype spellings used as catalog keys. Instance/spec dtype strings
# are normalised through this map so "f16"/"half" and "fp16" all resolve.
_DTYPE_ALIASES = {
    "f16": "fp16",
    "half": "fp16",
    "fp16": "fp16",
    "bf16": "bf16",
    "bfloat16": "bf16",
    "f32": "fp32",
    "float": "fp32",
    "fp32": "fp32",
    "fp8": "fp8e4m3",
    "fp8e4m3": "fp8e4m3",
    "bf8": "bf8e5m2",
    "bf8e5m2": "bf8e5m2",
    "fp6": "fp6e2m3",
    "fp6e2m3": "fp6e2m3",
    "bf6": "fp6e3m2",
    "fp6e3m2": "fp6e3m2",
    "fp4": "fp4e2m1",
    "fp4e2m1": "fp4e2m1",
    # Integer WMMA: "iu8"/"iu4" are the RDNA WMMA integer operand families
    # (signedness is an instruction operand, not the dtype); "i32" is the
    # integer accumulator. Scalar int spellings pass through for completeness.
    "iu8": "iu8",
    "iu4": "iu4",
    "i8": "i8",
    "int8": "i8",
    "i4": "i4",
    "int4": "i4",
    "i32": "i32",
    "int32": "i32",
}


def normalize_dtype(name: str) -> str:
    """Map a dtype spelling to its canonical catalog key."""
    key = name.strip().lower()
    return _DTYPE_ALIASES.get(key, key)
